### ISSUE TITLE
tpetra:  TpetraCore_CrsMatrix_UnitTests4_MPI_4 fails in cuda builds

### ISSUE THREAD

**Issue by maartenarnst** [trilinos/Trilinos#12898]
@trilinos/tpetra
@brian-kelley
@csiefer2 

### Description

We're seeing the test `TpetraCore_CrsMatrix_UnitTests4_MPI_4` fail in cuda builds. We use cuda 12.3, and we enable cusparse in our builds. 

The failures look like:

```
CrsMatrix_int_longlong_double_Tpetra_KokkosCompat_KokkosCudaWrapperNode_ApplyHelpers_UnitTest ...
Comparing exact == norm ... passed
Comparing exact == norm ... passed
did_we_batch = true == should_we_batch = true : passed
Comparing exact == norm ...
Error, relErr(exact[0],norm[0]) = relErr(0,3.8858e-11) = 0.999994 <= tol = 4.44089e-16: failed!

Error, relErr(exact[1],norm[1]) = relErr(0,3.88294e-11) = 0.999994 <= tol = 4.44089e-16: failed!
Comparing exact == norm ... passed
did_we_batch = true == should_we_batch = true : passed
Comparing exact == norm ...
```

The failures happen in the `ApplyHelpers` test. Here, spmv operations are computed in two different ways, first using a usual way, and then using a "batched" way. The test requires equality of the results, with a very restrictive tolerance basically up to machine precision.

It appears that something has changed the algorithm that is used in at least one of the two ways in which the spmv is computed. As a result, the result is still OK, but the test fails because the result is no longer the same up to machine precision.  

- https://github.com/trilinos/Trilinos/blob/f5169f5738f2b73bdcc393dbc8ee893e41c6f8ac/packages/tpetra/core/test/CrsMatrix/CrsMatrix_UnitTests4.cpp#L1135

I do not know why the failure shows up in our but seemingly not your Trilinos cuda tests.

The failures might have something to do with:
- https://github.com/trilinos/Trilinos/pull/12852

**Comment by brian-kelley** [comment]
@maartenarnst Could you say which CUDA version you're using? I didn't see this happen when I tested the KokkosKernels 4.3 changes, but I _did_ see this test fail when testing the non-default cuSPARSE spmv algorithm with CUDA 11.6.1 for the same reason.

The changes I merged never use this alternate algorithm though, so either there is some numerical differences between cuSPARSE versions, or it's nondeterministic behavior.

Once I replicate this, I'll figure out if 3.9e-11 is a reasonable error norm for the matrix/vector. If it is, we probably just need to change the tolerance or the way the exact and actual vectors are compared.

**Comment by maartenarnst** [comment]
Hi @brian-kelley. 

We use:
- cuda version 12.3.2
- cusparse version NV_LIBCUSPARSE_VERSION=12.2.0.103-1
- gcc 12.3.0

In fact, we build Trilinos and the dependencies in a docker image based on `nvidia/cuda:12.3.2-devel-ubuntu22.04`.

**Comment by brian-kelley** [comment]
Thanks @maartenarnst. I edited my comment because I saw you did in fact have the version right in the first line 🤦 

I did try to build this test on the two machines I know of with CUDA 12 - one has an insufficient driver, and the other (I think) has no MPI that's compatible. Instead I'll try to emulate the behavior by switching to the alternate algorithm ``CUSPARSE_SPMM_CSR_ALG2`` with CUDA 11.6.
